### DISCUSSION TITLE
Trouble doing something as simple as api.getNote(noteID)... Help?

### DISCUSSION THREAD

**Discussion by sottey** [TriliumNext/Trilium#4210] [Q&A] [answered]
I am tryiing to use a note's json contents as a data source and I am unable to get the return from my getNoteContents(noteID) function to be anything other than undefined. I am a long time developer, but new-ish to js.

**My function I am calling:**
```
async function getNoteContent(noteID, content) {
    return await api.runOnBackend((noteID) => {
        const becca = require('../becca/becca');
        const noteService = require('./notes');
        
        const note = becca.getNote(noteID);

        if (!note) {
            return [404, `Note ${noteID} has not been found`];
        }

        if (!note.isContentAvailable()) {
            return [400, `Note ${noteID} is not available for change`];
        }
        
        content = note.getContent();

    }, [noteID, content]);
}    

```

**How I am calling it** ( from inside a function defined as: `async function loadEvents()` ):

`var noteContent = await getNoteContent('TR2pVO3d7opK');`

Always returns "undefined". I have verified and reverified that:
- The ID is the correct id for the note
- The note is of type Code -> JSON
- The note has only valid JSON data in it

Thoughts?

Thank you!
Sean

**Comment by zerebos** [comment]
You are not `return`ing anything from the function on the backend so nothing gets returned to you. It seems you're trying to use a parameter as a return but that's only generally possible with things like objects that are pass by reference. And in this case since we are crossing contexts and everything is serialized, it isn't possible here at all.

So to fix it replace `content = note.getContent();` with `return note.getContent();`

**Or** if you're just after the content, no need to run it on the backend at all.

```js
const targetNote = await api.getNote('TR2pVO3d7opK');
const noteContent = await targetNote.getContent();
```

**Reply by zerebos** [reply]
I'm also curious as to why you're trying to do things with `becca` directly via `require` instead of making use of the [backend api](https://zadam.github.io/trilium/backend_api/BackendScriptApi.html)

**Reply by sottey** [reply]
I basically cobbled this together from code snippets on the web.

**Comment by sottey** [comment]
So if I replace all my nonsensical code with **this**:
```
            var noteContent = await getDataNoteContent();
            alert("noteContent: " + noteContent); 
```


**calling this**:
```
async function getDataNoteContent() {
    const targetNote = await api.getNote('TR2pVO3d7opK');
    const noteContent = await targetNote.getContent();
    return noteContent;
} 
```

The alert shows: 

> noteContent: undefined

Back where I started :-(

Thanks for your responses though, it is appreciated!
Sean

**Comment by zerebos** [comment]
Are you doing anything else in your script? I tested this myself with a valid note ID and it works just fine

![image](https://github.com/zadam/trilium/assets/6865942/17b7b0bb-133a-4d31-8e0a-d2116b12fa05)
![image](https://github.com/zadam/trilium/assets/6865942/da578b1a-7df4-4af0-b470-6bd48f82f297)


Are you on Trilium stable or beta?

**Reply by sottey** [reply]
Hmmm. Isolating like that is smart. I will try that now. I am basically integrating a complete javascript system into Trilium, so there is a lot going on outside of my code.

Thanks again!
Sean

**Reply by sottey** [reply]
Also, excellent question, I should have stated this in the beginning. I am running Trilium 0.61.5-beta. 

I wonder if there is an issue there.

![image](https://github.com/zadam/trilium/assets/8494620/7317e8e5-7a0a-466a-bbd6-70aed113cb92)

**Reply by sottey** [reply]
I also have had some issues regarding the reverse proxy which I believe to be sorted (running successfully on https via a custom domain) but that is another wrinkle.

I think, just to be sure, I will stand up another instance of Trilium with the stable build and see if that works for me

Thanks again,.
Sean

**Reply by sottey** [reply]
SO...

Here is my 0.61.5-beta instance:
![image](https://github.com/zadam/trilium/assets/8494620/5109aaa2-09da-4479-9525-2ed275a686f9)

and here is my stable instance:
![image](https://github.com/zadam/trilium/assets/8494620/1e40a1ee-52b5-4bb7-bedd-753f4a2b96ee)


SO, something in my contents OR something in the beta is causing issues.... The stable instance is a fresh docker instance. My beta instance has a TON of code and scripts and other stuff going on...

Thanks,
Sean

**Comment by sottey** [comment]
Ok, this is truly, really, the last post.

SO, I have my primary Trilium (lots of content, 0.61.5-beta) that I call "prod". I now have another instance, which is fresh, running the stable release which I call "stable". I have a third instance which is an empty 0.61.5-beta instance that I call "beta".

PROD: getContent() returns undefined
STABLE: getContent() returns the note contents
BETA: getContent() returns undefined

I am feeling comfortable, with all this testing, that there is an issue with getContent() in the 0.61.5-beta version. 

Process:

1. coded the work I have on my primary (PROD), and it failed, so I posted here.
2. Code failed with getContent returning undefined
3. Exported the note tree with my code in it.
4. Created a new instance with nothing in it using the stable release version
5. Imported that exported note tree zip
6. Code works just fine
7. Created a new instance with nothing in it using the beta release version
8. Imported that exported note tree zip
9. Code fails in the same way as in step 2, getContent() returns undefined.

Hope this all helps.

Thanks,
Sean

**Reply by zerebos** [reply, marked as answer]
You're right this appears to be not working in the beta. @zadam is `FNote.getContent()` no longer meant to work in the beta? I'm guessing no since it's not documented in the jsdoc and ultimately replaced with `getBlob` right?

For now, sottey, you can just do this:

```js
function getNoteContent(noteId) {
    return api.runOnBackend((id) => api.getNote(id).getContent(), [noteId]);
}

var noteContent = await getNoteContent("TR2pVO3d7opK");
alert("noteContent: " + noteContent); 
```

this method should work on both stable and the beta

or if you only care about the beta you can move to `getBlob` which returns a promise that resolves to the `FBlob` object that has a `content` member like this:

```js
async function getNoteContent(noteId) {
    const targetNote = await api.getNote(noteId);
    const noteBlob = await targetNote.getBlob();
    return noteBlob.content;
}

var noteContent = await getNoteContent("1Nq6mstCRxG2");
alert("noteContent: " + noteContent); 
```

**Reply by zadam** [reply]
0.61 contains a lot of changes for how content is stored/retrieved and this was one regression, thanks for reporting it. It will be fixed in the next release.

**Reply by zerebos** [reply]
Ah okay, I wasn't sure if it was intended or not, I'll try to make sure to report any other inconsistencies I find

**Comment by sottey** [comment]
Thanks again for all the help. I have moved back to the stable release for this. 

You are the best!

**Reply by zerebos** [reply]
No worries, you can stay on the beta if you like with that last snippet I gave. If you're using a lot of javascript you might be interested in the `@types` package I have for trilium frontend and backend apis
